app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt

```diff
@@ -31,34 +30,47 @@ class NotificationFetcher @Inject constructor(
         }
     }
 
+    /**
+     * Fetch new Mastodon Notifications and update the marker position.
+     *
+     * Here, "new" means "notifications with IDs newer than notifications the user has already
+     * seen."
+     *
+     * The "water mark" for Mastodon Notification IDs are stored in two places.
+     *
+     * - acccount.lastNotificationId -- the ID of the top-most notification when the user last
+     *   left the Notifications tab.
+     * - The Mastodon "marker" API -- the ID of the most recent notification fetched here.
+     *
+     * The user may have refreshed the "Notifications" tab and seen notifications newer than the
+     * ones that were last fetched here. So `lastNotificationId` takes precedence if it is greater
+     * than the marker.
+     */
     private fun fetchNotifications(account: AccountEntity): MutableList<Notification> {
         val authHeader = String.format("Bearer %s", account.accessToken)
-        // We fetch marker to not load/show notifications which user has already seen
-        val marker = fetchMarker(authHeader, account)
-        if (marker != null && account.lastNotificationId.isLessThan(marker.lastReadId)) {
-            account.lastNotificationId = marker.lastReadId
+
+        val minId = when (val marker = fetchMarker(authHeader, account)) {
+            null -> account.lastNotificationId.takeIf { it != "0" }
+            else -> if (marker.lastReadId > account.lastNotificationId) marker.lastReadId else account.lastNotificationId
         }
-        Log.d(TAG, "getting Notifications for " + account.fullName)
+
+        Log.d(TAG, "getting Notifications for ${account.fullName}, min_id: $minId")
+
         val notifications = mastodonApi.notificationsWithAuth(
             authHeader,
             account.domain,
-            account.lastNotificationId
+            minId
         ).blockingGet()
 
-        val newId = account.lastNotificationId
-        var newestId = ""
-        val result = mutableListOf<Notification>()
-        for (notification in notifications.reversed()) {
-            val currentId = notification.id
-            if (newestId.isLessThan(currentId)) {
-                newestId = currentId
-                account.lastNotificationId = currentId
-            }
-            if (newId.isLessThan(currentId)) {
-                result.add(notification)
-            }
-        }
-        return result
+        if (notifications.isEmpty()) return mutableListOf()
+
+        // Notifications are returned in order, most recent first. Save the newest notification ID
+        // in the marker.
+        val newMarkerId = notifications.first().id
+        Log.d(TAG, "updating notification marker to: $newMarkerId")
+        mastodonApi.updateMarkersWithAuth(authHeader, notificationsLastReadId = newMarkerId)
```

Comment:
This must also be done in the notification tab?

Reply:
No, because the notification tab (specifically, NotificationsViewModel) updates `account.lastNotificationId` when the user navigates away, and `account.lastNotificationId` is used in the algorithm earlier in this code for figuring out which notification to show.

The notifications tab should show an unbroken timeline of the user's notifications, unaffected by the Android Notifications that are generated. When the user returns to this tab their reading position (i.e., the notification they return to) should be the same one as when they left. That's why the notification tab has a separate "marker" for the user's position.

However Android Notifications do not need to show every single Mastodon Notification. If the user has already seen it in the notifications tab then they don't need to see an Android Notification about it as well. So the "marker" for the Android Notifications is set to whichever is larger of either the previous value, or the marker that the notification tab uses.

Reply:
Then we are talking of the same. (Wanted to avoid exactly that: get Android notifications for a previously seen Mastodon notification)

Reply:
> No, because the notification tab (specifically, NotificationsViewModel) updates `account.lastNotificationId` when the user navigates away

"navigates away" means I still get an Android notification about a notification I am looking at at the moment in the timeline?
Shouldn't it be "once loaded"?

Reply:
There's probably more to it than that. I just experimented with the GMail app, which has the following behaviours that I think we should imitate.

## Active GMail Android Notifications are cleared on open

- Have one or more GMail Android Notifications
- Open GMail
- The Android Notifications are cleared

Proposed Tusky behaviour -- clear all Tusky Android Notifications for the active account when the user navigates to the Notifications tab.

## No Android Notifications while GMail is open

- Open GMail
- From another e-mail app on a different device, send a message to the GMail account you have open on the phone
- GMail updates to show the new e-mail, but does not trigger an Android Notification

Proposed Tusky behaviour -- do not send an Android Notification while the Notifications tab is open. However, do update the marker if new notifications are fetched by the NotificationFetcher code.

Reply:
> Proposed Tusky behaviour -- clear all Tusky Android Notifications for the active account when the user navigates to the Notifications tab.

That should be the case: #2528

> Proposed Tusky behaviour -- do not send an Android Notification while the Notifications tab is open. However, do update the marker if new notifications are fetched by the NotificationFetcher code.

I see that falling in the category "update things without user interaction / in the background".
One aspect of that would be some information badges with "here is new content".
In the case of the notifications timeline it could update the list silently (and still show the "new stuff" badge on the tab icon).

Reply:
> > Proposed Tusky behaviour -- clear all Tusky Android Notifications for the active account when the user navigates to the Notifications tab.
> 
> That should be the case: #2528

That's not happening -- there's a bug somewhere.

I've got an emulator running with the code from this PR, and the active account has 5 Android Notifications (i.e., if I pull down the notification shade I see five notifications).

When I switch to the notification tab `NotificationHelper.clearNotificationsForActiveAccount()` *is* called.

But inside that function `account.getActiveNotifications()` returns `"[]"`, so the notifications are never cancelled.

I put breakpoints on every place in the code where `AccountEntity.activeNotifications` is modified, and launched a debug build. The breakpoints were never hit, so the list of active notifications is not being updated / fetched.

I'll keep digging.

Reply:
> > That should be the case: #2528
> 

> When I switch to the notification tab `NotificationHelper.clearNotificationsForActiveAccount()` _is_ called.
> 
> But inside that function `account.getActiveNotifications()` returns `"[]"`, so the notifications are never cancelled.


That sounds "basic" as if notification removal does not/never work for you?
It works fine for me (either pull or now push notifications).

However a quick look in the code I found the `NotificationClearBroadcastReceiver` which does exactly that: Only set the field to empty array but does not actually clear any notifications (?).
I don't remember however ever stumbling across this code or problem therein.

Reply:
From experimenting with the code it would clear the record of all the Android notifications if the user removed (e.g., swiped left) on a single notification.

That's obviously bad, so I've also fixed that in this PR -- instead of trying to keep track of the active notifications it's easier to get the active notifications when they're needed and iterate over them.

This also simplifies the code -- there's no need for the database column, marshalling data to/from JSON, etc.

Reply:
This ensures that notifications are cleared when the notifications tab is active.

It doesn't disable showing new notifications if they come in while the tab is open. That can get pushed to v23, I'm focused on trying to fix the known issues with v22 so we can get beta 3 out of the door and then hopefully a final release.